controllers/cruisecontroloperation_ttl_controller.go

```diff
@@ -0,0 +1,116 @@
+// Copyright © 2022 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/banzaicloud/koperator/api/v1alpha1"
+	banzaiv1alpha1 "github.com/banzaicloud/koperator/api/v1alpha1"
+)
+
+const (
+	defaultRequeueCruiseControlOperatonTTL = 30
+)
+
+// CruiseControlOperationTTLReconciler reconciles CruiseControlOperation custom resources
+type CruiseControlOperationTTLReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations,verbs=get;list;watch;create;update;patch;delete;deletecollection
+// +kubebuilder:rbac:groups=kafka.banzaicloud.io,resources=cruisecontroloperations/status,verbs=get
+
+//nolint:gocyclo
+func (r *CruiseControlOperationTTLReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContextOrDiscard(ctx)
+	log.V(1).Info("reconciling CruiseControlOperation custom resources")
+
+	ccOperation := &banzaiv1alpha1.CruiseControlOperation{}
+	if err := r.Get(ctx, client.ObjectKey{Name: req.Name, Namespace: req.Namespace}, ccOperation); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			return reconciled()
+		}
+		// Error reading the object - requeue the request.
+		return requeueWithError(log, err.Error(), err)
+	}
+
+	if ccOperation.GetTTLSecondsAfterFinished() == nil || ccOperation.GetCurrentTask().Finished == nil {
+		return reconciled()
+	}
+
+	operationTTL := time.Duration(*ccOperation.GetTTLSecondsAfterFinished())
```

Comment:
**Required**: We **MUST** handle the case where the TTL is nil, because it is a valid and often use-case.

```suggestion
	operationTTL := defaultRequeueCruiseControlOperatonTTL * time.Seconds // todo(Márton): if you change the type of the default from int to time.Duration, then upgrade this one as well to drop the conversion.

	setTTLSeconds := ccOperation.GetTTLSecondsAfterFinished()
	if setTTLSeconds != nil {
		operationTTL = *setTTLSeconds * time.Seconds
	}
```

Reply:
I have fixed this bug in a commit later.